### DISCUSSION TITLE
Did not recv blocks

### DISCUSSION THREAD

**Discussion by jnikoletich** [blockworks-foundation/lite-rpc#315] [General]
Running the latest version, I am getting the follow error on startup:

```
thread 'main' panicked at lite-rpc/src/main.rs:59:5:
Did  not recv blocks
```

I have confirmed my RPC and WS connection.

Here is a full stacktrace if needed:

```
stack backtrace:
   0:     0x55be8e1c99bc - std::backtrace_rs::backtrace::libunwind::trace::he43a6a3949163f8c
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55be8e1c99bc - std::backtrace_rs::backtrace::trace_unsynchronized::h50db52ca99f692e7
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55be8e1c99bc - std::sys_common::backtrace::_print_fmt::hd37d595f2ceb2d3c
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x55be8e1c99bc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h678bbcf9da6d7d75
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x55be8e1f625c - core::fmt::rt::Argument::fmt::h3a159adc080a6fc9
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/fmt/rt.rs:138:9
   5:     0x55be8e1f625c - core::fmt::write::hb8eaf5a8e45a738e
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/fmt/mod.rs:1094:21
   6:     0x55be8e1c61ce - std::io::Write::write_fmt::h9663fe36b2ee08f9
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/io/mod.rs:1714:15
   7:     0x55be8e1c97a4 - std::sys_common::backtrace::_print::hcd4834796ee88ad2
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x55be8e1c97a4 - std::sys_common::backtrace::print::h1360e9450e4f922a
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x55be8e1cb093 - std::panicking::default_hook::{{closure}}::h2609fa95cd5ab1f4
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:270:22
  10:     0x55be8e1cadac - std::panicking::default_hook::h6d75f5747cab6e8d
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:290:9
  11:     0x55be8e1cb619 - std::panicking::rust_panic_with_hook::h57e78470c47c84de
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:707:13
  12:     0x55be8e1cb4d1 - std::panicking::begin_panic_handler::{{closure}}::h3dfd2453cf356ecb
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:597:13
  13:     0x55be8e1c9ee6 - std::sys_common::backtrace::__rust_end_short_backtrace::hdb177d43678e4d7e
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/sys_common/backtrace.rs:170:18
  14:     0x55be8e1cb262 - rust_begin_unwind
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:595:5
  15:     0x55be8d66b823 - core::panicking::panic_fmt::hd1e971d8d7c78e0e
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:67:14
  16:     0x55be8d82838c - lite_rpc::start_lite_rpc::{{closure}}::h21db0d2276f4b1d5
  17:     0x55be8d81d186 - <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll::h2aa59afc3fb5bb56
  18:     0x55be8d873104 - tokio::runtime::park::CachedParkThread::block_on::hde28a404e03cf620
  19:     0x55be8d710e77 - lite_rpc::main::h7a705928c087e212
  20:     0x55be8d8107f3 - std::sys_common::backtrace::__rust_begin_short_backtrace::h3dd76d4f7f9b812b
  21:     0x55be8d7615cd - std::rt::lang_start::{{closure}}::h8e3329752f705dc6
  22:     0x55be8e1bf23b - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hbcc4f8a3f5ada78c
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:284:13
  23:     0x55be8e1bf23b - std::panicking::try::do_call::he5f117a9e13dadde
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:502:40
  24:     0x55be8e1bf23b - std::panicking::try::h2f3af9afce3a0443
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:466:19
  25:     0x55be8e1bf23b - std::panic::catch_unwind::h6d6c387f38ef05ea
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panic.rs:142:14
  26:     0x55be8e1bf23b - std::rt::lang_start_internal::{{closure}}::h6ca09d5905711415
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/rt.rs:148:48
  27:     0x55be8e1bf23b - std::panicking::try::do_call::ha9fd18ea06654a4b
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:502:40
  28:     0x55be8e1bf23b - std::panicking::try::hda5c2a4432362341
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:466:19
  29:     0x55be8e1bf23b - std::panic::catch_unwind::h440f731b142bc235
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panic.rs:142:14
  30:     0x55be8e1bf23b - std::rt::lang_start_internal::hc0b4e50f058f62ce
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/rt.rs:148:20
  31:     0x55be8d711815 - main
  32:     0x7f8ce807d083 - __libc_start_main
  33:     0x55be8d66beee - _start
  34:                0x0 - <unknown>

```

**Comment by godmodegalactus** [comment]
Sorry for delays,
Which RPC are you using ?

**Comment by leonardo3dp** [comment]
why do I need to set an rpc server? wouldn't lite-rpc connect to the validator like normal rpc do?

**Comment by aniketfuryrocks** [comment]
It defaults to `solana-local-validator` which is at address `http://0.0.0.0:8899`.
In order to use the mainnet-beta. Please provide the url as an argument.

**Comment by godmodegalactus** [comment]
> why do I need to set an rpc server? wouldn't lite-rpc connect to the validator like normal rpc do?

Lite-rpc aims to provide scalability to RPC infrastructure. It will depend on one or more Solana RPC to do load balancing on several functionality. Lite-RPC cannot replay blocks to confirm and finalize itself, but it takes block stream from a RPC and implements sending and confirming transactions based on it.
You can also provide tritons geyser endpoints which are faster and efficient to get a block stream.

**Comment by leonardo3dp** [comment]
it needs an rpc to get the information! Is the transaction sent to the RPC or directly to the validators?

**Comment by de-ltd** [comment]
I would presume the RPC connections are just for the data the lite-rpc needs in order to function - I would also presume it then proceeds to send transactions directly to the validators.

But who knows honestly, not sure why everything solana related has such poor documentation.

**Comment by godmodegalactus** [comment]
> it needs an rpc to get the information! Is the transaction sent to the RPC or directly to the validators?

Yes it is sent directly to the validators. And then confirmation is done over data that comes from the rpc.

**Comment by godmodegalactus** [comment]
> I would presume the RPC connections are just for the data the lite-rpc needs in order to function - I would also presume it then proceeds to send transactions directly to the validators.
> 
> But who knows honestly, not sure why everything solana related has such poor documentation.

Yes transactions are sent directly to the validators.
Thanks for the feedback, we will start working more on the documentation part.

**Comment by de-ltd** [comment]
Good thanks - so what is the ideal setup here? Im trying to use my Helius RPC url to test the client - I see the default is to connect to "solana-validator" locally but what is that exactly.. a full node? 

Also how are configs actually passed in the run command:
cargo run --release --config RPC_ADDR="https://mainnet.helius-rpc.com/?api-key=blabla"

doesnt seem to work as it doesn't read the string

**Comment by aniketfuryrocks** [comment]
Not exactly.
you need to edit the config file. Earlier we had command line arguments but it was recently removed.
Steps:
1. Edit the config with desired values. Please refer to the README for all the key-value pairs.
2. Run the lite rpc `cargo run --release -- -c ./<your-config-name>.json`

Other wise you can also use the env variables. Please refer to the README for all the key-values.
In your case

```bash
env RPC_ADDR="https://mainnet.helius-rpc.com/?api-key=blabla" cargo run --release # -- pass any arguments if desired
```

Note:
Config preference.

env > config

You can find the example config in the `config.example.json`

```json
{
  "rpc_addr": "http://0.0.0.0:8899",
  "ws_addr": "ws://0.0.0.0:8900",
  "lite_rpc_http_addr": "[::]:8890",
  "lite_rpc_ws_addr": "[::]:8891",
  "fanout_size": 18,
  "identity_keypair": null,
  "prometheus_addr": "[::]:9091",
  "maximum_retries_per_tx": 40,
  "transaction_retry_after_secs": 3,
  "quic_proxy_addr": null,
  "use_grpc": false,
  "calculate_leader_schedule_form_geyser": false,
  "grpc_addr": "http://127.0.0.0:10000",
  "grpc_x_token": null,
  "postgres": {
    "pg_config": "your_postgres_config",
    "ssl": {
      "ca_pem_b64": "your_base64_encoded_ca_pem",
      "client_pks_b64": "your_base64_encoded_client_pks",
      "client_pks_pass": "your_client_pks_password"
    }
  }
}
```

**Comment by de-ltd** [comment]
cargo run --release -- -c ./<your-config-name>.json

Are you sure this is correct: error: unexpected argument '-c' found

**Comment by aniketfuryrocks** [comment]
which branch are you on ? I just tried it 

```bash
cargo run --release -- -c ./config.example.json 
```

Consider pulling the branch if you are on main

**Comment by de-ltd** [comment]
Which branch - i did think that as i was using the release version so i redownloaded the main branch just now but still no look, i see you are working on the production branch shall i try that?

**Comment by aniketfuryrocks** [comment]
I am using the main branch. I'll simplify this for you.

Steps:
1. make sure cargo is installed. run `cargo --version` to verify.
2. Install lite-rpc using `cargo` globally.

```bash
cargo --version # Mine is cargo 1.73.0 (9c4383fb5 2023-08-26)
cargo install --git https://github.com/blockworks-foundation/lite-rpc lite-rpc --locked
```

Run with logs

```
env RUST_LOG=info lite-rpc
```

Attach config

```
env RUST_LOG=info lite-rpc -c <CONFIG-PATH>
```

Note:
In case there is an issue resolving the command `lite-rpc`. Consider checking if cargo bin is in your path.

**Comment by de-ltd** [comment]
Thanks for that, I have indeed got that working now.

```
env RUST_LOG=info lite-rpc -c ./config.example.json
2024-02-01T16:17:12.852761Z  INFO lite_rpc: Use RPC address: https://mainnet.helius-rpc.com/?api-key=blabla    
2024-02-01T16:17:12.852789Z  INFO lite_rpc: grpc_sources:[GrpcSource { addr: "http://127.0.0.0:10000", x_token: None }]    
2024-02-01T16:17:12.852840Z  INFO lite_rpc: Creating RPC poll subscription...    
2024-02-01T16:17:12.852923Z  INFO lite_rpc: Waiting for first finalized block...    
2024-02-01T16:17:12.853024Z  INFO solana_lite_rpc_cluster_endpoints::rpc_polling::poll_blocks: block listener started    
2024-02-01T16:17:15.887063Z  INFO lite_rpc: Got finalized block: 245506271    
2024-02-01T16:17:16.051703Z ERROR lite_rpc: Services quit unexpectedly Err(invalid connection string: unexpected EOF
```


I used the example config json file - i only changed the rpc_addr to the helius one and the ws_addr to the helius web socket address.

Do you know what connection string this is complaining about? Thanks.

**Reply by godmodegalactus** [reply]
I think check with helius, you can also use grpc with helius. Usually it is helius endpoint as grpc_addr and your API key as grpc_x_token.

**Reply by godmodegalactus** [reply]
Update config.example.json accordingly as you are using mainnet you should point everything to mainnet.

**Reply by dschrengost** [reply]
did you ever figure this out?  I'm getting the exact same error "unexpected EOF"

**Reply by aniketfuryrocks** [reply]
use `jq` to verify if the json file is valid. 

```
jq -e <path-to-file>
```

**Reply by DYSIM** [reply]
I reached the same stage, and looks like everything connected as I can see the logs from alchemy dashboard. problem is alchemy is returning errors like -

{
jsonrpc:"2.0",
error:{
code:-32009,
message:"Slot 258875925 was skipped, or missing in long-term storage"
},
id:47
}

there are many of these, so it looks like lite-rpc is giving strange parameter requests

**Comment by brandonros** [comment]
What is the benefit of running this and making it point to Helius? I would have thought (misunderstood) this was a replacement/substitute for needing Helius/equivalent.

**Reply by godmodegalactus** [reply]
The goal is to create a lite weight app which can send transactions directly to the cluster. It is not aimed to replace traditional RPC but aims to improve transaction delivery. It is mainly beneficial for RPC providers / validators to submit transactions to the cluster with minimum latency.

**Reply by brandonros** [reply]
That is what I thought, but the fact that it can't seem to run without the dependency of an RPC is confusing.

For example, https://github.com/lmvdz/tpu-client

As an end user, you can use this to "skip" and go straight to the leader. That has pros/cons for sure. I would've expected similar to be possible here, am I misunderstanding?
